core/dbt/graph/selector_methods.py

```diff
@@ -577,6 +578,39 @@ def search(
             if node in matches:
                 yield node
 
+class SourceFreshSelectorMethod(SelectorMethod): #TODO: this requires SelectorMethod to have current_state as an argument. currently, this works because it's all hard-coded
+    def search(
+        self, included_nodes: Set[UniqueId], selector: str
+    ) -> Iterator[UniqueId]:
+
+        if self.previous_state is None or \
+           self.previous_state.current_sources is None or \
+           self.previous_state.archive_sources is None:
+            raise InternalException(
+                'No previous state comparison freshness results in sources.json'
+            )
```

Comment:
If we're missing either source result (previous or current), is it better to raise this explicit error, or to return all sources as selected?

This question could apply equally to `state:` and `result:` selection; we might end up with different answers for each, though I'd obviously prefer consistency.